### PR TITLE
Safe FS writing

### PR DESCRIPTION
**Pull request by Zarel**
Writing program state to a file is fairly hard to do safely, especially with Node's async FS writing. PS previously reimplemented the code necessary to do it safely basically everywhere it was needed. FS().writeUpdate now consolidates that code so anyone can easily safely update a file.

Would like code review on this.

#4110 should be rebased on this.

### CODE REVIEW

File: rooms.js [hunk 1]
```diff
@@ -550,6 +503,20 @@ class GlobalRoom extends BasicRoom {
...
+	writeChatRoomData() {
+		FS('config/chatrooms.json').writeUpdate(() => (
+			JSON.stringify(this.chatRoomDataList)
+				.replace(/\{"title":/g, '\n{"title":')
+				.replace(/\]$/, '\n]')
+		));
+	}
+
+	writeNumRooms() {
+		FS('config/chatrooms.json').writeUpdate(() => (
```
Comment (sirDonovan) [inline rooms.js]: The file path here should be <code>'logs/lastbattle.txt'</code>.

**Comment by Zarel** [comment]
The punishments.js change is easier to read ignoring whitespace: https://github.com/Zarel/Pokemon-Showdown/pull/4145/files?w=1

File: fs.js [hunk 2]
```diff
@@ -285,4 +336,11 @@ function getFs(path) {
 	return new FSPath(path);
 }
 
-module.exports = getFs;
+const FS = Object.assign(getFs, {
+	/**
+	 * @type {Map<string, [Promise, (() => string | Buffer)?, Object]>}
+	 */
+	pendingUpdates: new Map(),
```
Comment (Morfent) [inline fs.js]: Is it important to be able to access this from outside the module? I would've just made it a constant and handled the logic for it from a `pendingUpdate` getter/setter on `FSPath`. I'm a bit wary of needing to mutate `getFs`, but I'll admit it doesn't make any real difference whether it is or not.

**Comment by Morfent** [comment]
This is a bit off-track, but would it be a good idea to cache `FSPath` instances by their absolute path from `FS`? This would optimize code making multiple calls to `FS`, but more importantly would allow `FSPath` to be able to be able to keep state without needing to rewrite code in other modules to keep track of their own `FSPath` instances. `FSPath#writeUpdate` wouldn't need `FS.pendingUpdates` in order to function.

Apart from that, I don't have anything to add. This looks good as is.

File: fs.js (hunk 2 above)
Comment (Zarel) [reply fs.js]: PS convention is to leave all variables public and globally accessible, for ease of manual hotpatching. We couldn't possibly hotpatch `writeUpdate` if it referred to a private variable, etc.

As for mutating `getFs`, literally all our libraries work this way. `Users`, `Rooms`, `Ladders` are all functions with properties. It's a bit weird, but it makes for a very convenient API.

**Comment by Zarel** [comment]
I really don't like the idea of caching `FSPath` instances. Memory is limited, and the cost of creating new objects is negligible for how often we do it.

**Pull request merged**